src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -36,13 +38,57 @@ internal abstract partial class AbstractDiagnosticsTaggerProvider<TTag> : Asynch
     {
         private readonly IDiagnosticService _diagnosticService;
 
+        /// <summary>
+        /// Keep track of the ITextSnapshot for the open Document that was used when diagnostics were
+        /// produced for it.  We need that because the DiagnoticService does not keep track of this
+        /// snapshot (so as to not hold onto a lot of memory), which means when we query it for 
+        /// diagnostics, we don't know how to map the span of the diagnostic to the current snapshot
+        /// we're tagging.
+        /// </summary>
+        private static readonly ConditionalWeakTable<object, ITextSnapshot> _diagnosticIdToTextSnapshot = 
```

Comment:
A possible perf optimization here is to hold onto ITextVersion instead. Your .ToSnapshotSpan().TranslateTo() then would become a call to [Tracking.TrackSpanForwardInTime](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.tracking.trackspanforwardintime).ToSnapshotSpan() to the current version. Should be pretty easy to change.

(Versions are cheaper than snapshots in that they don't hold onto the actual text structure, but just the deltas. A snapshot always exposes it's version from the .Version property.)

Reply:
> to the current version. Should be pretty easy to change.

sounds good.  However, still not sure how i validate that the snapshots are from teh same buffer.  i.e. that the open/close/open scenario doesn't cause problems.

Reply:
Ok.  I couldn't get this to work.  The issue is that diagnostics store their information as line/cols.  Which means i need the actual original snapshot in order to map that information back to a reasonable span in roslyn coordinates.  I could attempt to store an ITextImage.  However, then i also need a way to map that to a SourceText (since that's the currency that DiagnosticData expects).  I could do all the work to provide a SourceText around an ITextImage, but this seems like a lot of extra gunk just to get this work.

Thoughts @jasonmalinowski

Reply:
Oh, that's lame. Why don't we just hold onto it in regular span form to start? Was it just to work around issues here that we're now making better?

Reply:
That's a good question.  @heejaechang Do you know why diagnostics are line/column instead of just TextSpan?

note: it may be because of external diagnostics where some tool is reporting the problem...  but i don't know for certain.

Reply:
DiagnosticData has line/column data if it came from build. but if it came from live, it will have TextSpan.

the way you do GetExistingOrCalculatedTextSpan should return TextSpan if it is already there or create one if it is not.

I dont calculate text span from build error right away because it requires reading in Text and a lot of time, those errors stay in error list (where it doesn't need span) and never shown to users in editor (which require span)

Reply:
Right.  But since these errors may have come from build, and not from live, we can't assume there will be TextSpans for them, and we have to provide the SourceText so that the diagnostic data can be mapped to a span.  Correct?

Reply:
yep